### PR TITLE
Exception notifications to user + Add LoadingWindow class for improved processing feedback

### PR DESCRIPTION
**Pull request by ItsSimko**
## Summary by Sourcery

Add a LoadingWindow class to enhance user feedback during processing by replacing manual popups with an animated window. Implement exception handling to notify users of errors during audio processing and transcription.

New Features:
- Introduce a LoadingWindow class to provide users with animated feedback during processing tasks.

Enhancements:
- Replace manual popup creation with the new LoadingWindow class for a more consistent and user-friendly processing feedback experience.
- Display exceptions to the user if issues processing audio with local whisper or remote whisper. Let them know why it failed...

## Summary by Sourcery

Add a LoadingWindow class to enhance user feedback during processing by replacing manual popups with an animated window. Implement exception handling to notify users of errors during audio processing and transcription.

New Features:
- Introduce a LoadingWindow class to provide users with animated feedback during processing tasks.

Bug Fixes:
- Implement exception handling to notify users of errors during audio processing and transcription.

Enhancements:
- Replace manual popup creation with the new LoadingWindow class for a more consistent and user-friendly processing feedback experience.

### CODE REVIEW

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: src/FreeScribe.client/UI/LoadingWindow.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        This method is called automatically by __init__ and shouldn't
+        typically be called directly.
+        """
+        current_text = self.popup_label.cget("text")
+        base_text = "Processing Audio"
+        
+        if current_text.endswith("..."):
+            self.popup_label.config(text=base_text)
+        else:
+            self.popup_label.config(text=current_text + ".")
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/LoadingWindow.py:95]: Not needed removed this section.

File: src/FreeScribe.client/UI/LoadingWindow.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        #Not Resizable
+        self.popup.resizable(False, False)
+        
+        # Disable closing of the popup manually
+        self.popup.protocol("WM_DELETE_WINDOW", lambda: None)
+        
+        # Start the animation
+        self.animate_text()
+    
+    def animate_text(self):
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/LoadingWindow.py]: https://github.com/ClinicianFOCUS/FreeScribe/pull/56/commits/15c36bae892ba46c4095d4d2cef6aa675cb6edc4

Changed.

File: src/FreeScribe.client/UI/LoadingWindow.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        self.popup = tk.Toplevel(parent)
+        self.popup.title(title)
+        self.popup.geometry("200x100")
+
+        if parent:
+            # Center the popup window on the parent window
+            parent.update_idletasks()
+            x = parent.winfo_x() + (parent.winfo_width() - self.popup.winfo_reqwidth()) // 2
+            y = parent.winfo_y() + (parent.winfo_height() - self.popup.winfo_reqheight()) // 2
+            self.popup.geometry("+{}+{}".format(x, y))
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/LoadingWindow.py]: https://github.com/ClinicianFOCUS/FreeScribe/pull/56/commits/f75aedd03cdc19d5318c01e920a73ac569b17deb

changed

File: src/FreeScribe.client/client.py [hunk 4]
```diff
@@ -365,37 +346,51 @@ def send_audio_to_server():
...
+        try:
+            # Load the specified Whisper model
+            model_name = app_settings.editable_settings["Whisper Model"].strip()
+            model = whisper.load_model(model_name)
 
-        # Load the specified Whisper model
-        model_name = app_settings.editable_settings["Whisper Model"].strip()
-        model = whisper.load_model(model_name)
+            # Determine the file to send for transcription
+            file_to_send = uploaded_file_path if uploaded_file_path else 'recording.wav'
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py]: https://github.com/ClinicianFOCUS/FreeScribe/pull/56/commits/e261361b525818d197110c054160e3b5b95f611f

Done

File: src/FreeScribe.client/UI/LoadingWindow.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        dots to create an ellipsis animation effect. The animation continues
+        until the popup is destroyed.
+        
+        Note
+        ----
+        This method is called automatically by __init__ and shouldn't
+        typically be called directly.
+        """
+        current_text = self.popup_label.cget("text")
+        base_text = "Processing Audio"
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/LoadingWindow.py]: Not needed removed.

File: src/FreeScribe.client/client.py [hunk 6]
```diff
@@ -423,24 +418,38 @@ def send_audio_to_server():
...
-                response = requests.post(app_settings.editable_settings["Whisper Endpoint"], headers=headers, files=files, verify=False)
-            else:
-                # Send the request with the audio file and headers/authorization
-                response = requests.post(app_settings.editable_settings["Whisper Endpoint"], headers=headers, files=files)
-            
-            # On successful response (status code 200)
-            if response.status_code == 200:
-                # Update the UI with the transcribed text
-                transcribed_text = response.json()['text']
+            response = -1
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py]: https://github.com/ClinicianFOCUS/FreeScribe/pull/56/commits/432c0094aa4514eb9d440b503e562ceddc2326fe

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: src/FreeScribe.client/client.py [hunk 7]
```diff
@@ -365,37 +346,51 @@ def send_audio_to_server():
...
-        file_to_send = uploaded_file_path if uploaded_file_path else 'recording.wav'
-        uploaded_file_path = None
+            # Transcribe the audio file using the loaded model
+            result = model.transcribe(file_to_send)
+            transcribed_text = result["text"]
 
-        # Transcribe the audio file using the loaded model
-        result = model.transcribe(file_to_send)
-        transcribed_text = result["text"]
+            # Update the user input widget with the transcribed text
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:374]: One day during a refactor post release

File: src/FreeScribe.client/client.py [hunk 8]
```diff
@@ -365,37 +346,51 @@ def send_audio_to_server():
...
-
         # Configure the user input widget to be editable and clear its content
         user_input.scrolled_text.configure(state='normal')
         user_input.scrolled_text.delete("1.0", tk.END)
 
         # Display a message indicating that audio to text processing is in progress
         user_input.scrolled_text.insert(tk.END, "Audio to Text Processing...Please Wait")
+        try:
+            # Load the specified Whisper model
+            model_name = app_settings.editable_settings["Whisper Model"].strip()
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:363]: Again refactor post release...

File: src/FreeScribe.client/client.py [hunk 9]
```diff
@@ -365,37 +346,51 @@ def send_audio_to_server():
 
     global uploaded_file_path
 
+    loading_window = LoadingWindow(root, "Processing Audio", "Processing Audio. Please wait.")
+
     # Check if Local Whisper is enabled in the editable settings
     if app_settings.editable_settings["Local Whisper"] == True:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:352]: #61

File: src/FreeScribe.client/UI/LoadingWindow.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+        if parent:
+            # Center the popup window on the parent window
+            parent.update_idletasks()
+            x = parent.winfo_x() + (parent.winfo_width() - self.popup.winfo_reqwidth()) // 2
+            y = parent.winfo_y() + (parent.winfo_height() - self.popup.winfo_reqheight()) // 2
+            self.popup.geometry(f"+{x}+{y}")
+
+            #disable parent so we can prcoess with out anything else
+            parent.attributes('-disabled', True)
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/LoadingWindow.py]: https://github.com/ClinicianFOCUS/FreeScribe/pull/56/commits/d49cf938a79da1134ee0970fffa0b2148148c658

File: src/FreeScribe.client/client.py [hunk 11]
```diff
@@ -423,24 +418,37 @@ def send_audio_to_server():
...
-                # Update the UI with the transcribed text
-                transcribed_text = response.json()['text']
+            try:
+                # Check for SSL and self-signed certificate settings
+                if str(app_settings.SSL_ENABLE) == "1" and str(app_settings.SSL_SELFCERT) == "1":
+                    # Send the request without verifying the SSL certificate
+                    response = requests.post(app_settings.editable_settings["Whisper Endpoint"], headers=headers, files=files, verify=False)
+                else:
+                    # Send the request with the audio file and headers/authorization
+                    response = requests.post(app_settings.editable_settings["Whisper Endpoint"], headers=headers, files=files)
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py:428]: Fixed earlier

File: src/FreeScribe.client/client.py [hunk 12]
```diff
@@ -599,41 +598,48 @@ def show_edit_transcription_popup(formatted_message):
...
+                        post_processed_note = send_text_to_chatgpt(f"{app_settings.editable_settings['Post-Processing']}\nNotes:{medical_note}")
+                        update_gui_with_response(post_processed_note)
+                    else:
+                        update_gui_with_response(medical_note)
+            else: # do not generate note just send text directly to AI 
+                ai_response = send_text_to_chatgpt(edited_text)
+                update_gui_with_response(ai_response)
+        except Exception as e:
+            #Logg
+            #TODO: Implement proper logging to system event logger
```
Comment (yingbull) [inline src/FreeScribe.client/client.py:638]: I'll allow it here but better to just make an issue to track.

**Review by yingbull — APPROVED** [review]

**Pull request merged**